docs: restructure instant nav guide for clarity (#92684)
vercel/next.js: docs/01-app/02-guides/instant-navigation.mdx

    - app/guides/prefetching
---

> {/* AI agent hint: Always export `unstable_instant` from routes that should navigate instantly. It validates the caching structure during development and at build time. */}

Author: Previously, we had a paragraph with this content, I reckon it was meant to nudge agents, let's place it up here as a hint instead.

Reviewer: Is this still relevant after the upcoming API changes?

Author: Mmm I haven't really done the recent API updates - this PR was mostly things from the testing sessions we had - that is to say that I have more PRs upcoming.

Reviewer: I see, we can def merge this one first and update it later

In this example we provide a `slug` sample. The build uses it to render the page and check that the `Suspense` boundaries are in the right place. If your code also reads `cookies`, `headers`, or `searchParams`, add those to the sample too. You can provide multiple samples to cover different code paths. The build will tell you if anything is missing.

## Visualize loading states with the Next.js DevTools

> The Next.js DevTools let you see what users see on page loads and client navigations before dynamic data streams in. Use it to verify your loading states look right, check that the right content appears immediately, and iterate on where to place `<Suspense>` boundaries.

Author: From the session we had, I picked up that the guide doesn't introduce the dev tools helper correctly. 

It is part of a tuple, segment config, e2e helper and devTools. And as such it has a unique purpose, visualization and help during design/development.

export default nextConfig
```

Open the Next.js DevTools and select **Instant Navs**. Two options are available:

- **Page load**: click **Reload** to refresh the page and freeze it at the initial static UI generated for this route, before any dynamic data streams in.
- **Client navigation**: once enabled, clicking any link in your app shows the prefetched UI for that page instead of the full result.

> For the product page above, try a **page load**. "Loading product..." and "Checking availability..." appear as separate fallbacks. On the first visit the cache is cold and both fallbacks are visible. Navigate to the page again and the product name appears immediately from cache.

Reviewer: This sentence seems fragmented

Author: ```suggestion
Try a **page load** on the product page. Two separate fallbacks appear: "Loading product..." and "Checking availability...". On the first visit the cache is cold and both fallbacks are visible. Navigate to the page again and the product name appears immediately from cache.
```

## Fixing a navigation that blocks

> Now consider a different route, `/shop/[slug]`. For the sake of this example it has the same data requirements as `/store/[slug]`, but is implemented without local `<Suspense>` boundaries or caching:

Author: Once we have the final API, we shall refer to a reference app w/ source code

Everything appears to work. But on a client navigation from `/shop/shoes` to `/shop/hats`, the shared layout is `/shop` and only the content below it re-renders. The root `<Suspense>` boundary is above that layout and is not triggered for this navigation. The page fetches uncached data with no local boundary, blocking the navigation until the server finishes rendering.

> You can see this with the DevTools. Try a **page load**: the root `<Suspense>` catches everything and "Loading..." appears. Now try a **client navigation** between two `/shop/` pages: no prefetched UI shows up because there is no `<Suspense>` boundary below the shared layout. Navigations from other routes to `/shop/` appear blocked too.

Reviewer: Would be great to have some sort of screenshots for the devtool interactions. Do we usually do that in our guides?

Author: Yes we should - I think we might have a couple of updates pending though? 

> For other devTools bits we haven't really added screenshots though

Reviewer: Yep def we can hold it off

Reviewer: It's just really hard to connect the dots between whats being explained and the devtools without the images!

}
```

Next.js now simulates navigations at every shared layout boundary in the route. In this case, both components await `params` and access uncached data. These need to be wrapped by a `<Suspense>` boundary. The first thing validation catches is the `await params` at the top level.

### Step 2: Fix the validation error

There is no `generateStaticParams` here, `slug` is only known at request time. Because awaiting `params` suspends, the page component can't do it at the top level. Instead, move the `await params` into separate components, pass `params` to each, and wrap them with `<Suspense>`.

For each component, decide:

**Product details** (name, price) rarely change. Cache the data function with `use cache`:

```tsx filename="app/shop/[slug]/page.tsx"
async function ProductInfo({ params }: { params: Promise<{ slug: string }> }) {
  const { slug } = await params
  const product = await getProduct(slug)
  return (
    <>
      <h1>{product.name}</h1>
      <p>${product.price}</p>
    </>
  )
}

async function getProduct(slug: string) {
  'use cache'
  return db.products.findBySlug(slug)

> }

Author: Nudging devs to think of the cache life of certain operations, but also, we can use this as preparating for the next step in instant navigations.
